pilot/pkg/xds/discovery.go

```diff
@@ -252,7 +252,8 @@ func (s *DiscoveryServer) dropCacheForRequest(req *model.PushRequest) {
 
 // Push is called to push changes on config updates using ADS.
 func (s *DiscoveryServer) Push(req *model.PushRequest) {
-	if !req.Full {
+	// If it is headless service endpoint update, we need to re-init service registry to prevent generating stale nds.
+	if !req.Full && !model.HasConfigsOfKind(req.ConfigsUpdated, kind.DNSName) {
```

Comment:
I am not sure this addresses my comment. `Full=true` means "Do not recompute push context".

I was suggesting we make it Full=true, but then in LDS/RDS we can skip if only DNSName is changed (add to skippedLdsConfigs, etc)